### ISSUE TITLE
Cannot use editor without content being wrapped in Paragraphs

### ISSUE THREAD

**Issue by theefer** [ProseMirror/prosemirror#344]
I haven't been able to find a way to edit rich text in ProseMirror without wrapping it in a Paragraph node. Is this a known limitation or a feature?

For example, passing in the following document into ProseMirror (v0.6.1):

``` js
const schemaSpec = new SchemaSpec({
  text: Text,
  paragraph: Paragraph,
  doc: Doc
});
const contentDoc = customSchema.node('doc', {}, [
  customSchema.text('foo bar')
]);
```

displays fine in the DOM, but when focused, throws the following errors:

```
Uncaught TypeError: Cannot read property 'node' of undefined
MouseDown (input.js:277)
handlers.mousedown (input.js:263)
(anonymous function) (input.js:63)

Uncaught TypeError: Cannot read property 'head' of undefined
readFromDOM (selection.js:140)
poller (selection.js:82)
```

The only way I have found to fix the error is to wrap the Text node into a Paragraph:

``` js
const contentDoc = customSchema.node('doc', {}, [
  customSchema.node('paragraph', {}, [
    customSchema.text('foo bar')
  ])
]);
```

(I tried wrapping the Text node into an Inline, Block or Textblock node, but those yielded errors either when creating the doc, or when passing it to PM.)

The reason for trying to do away with Paragraphs is that I am trying to use ProseMirror for a single-line rich-text field (e.g. more like a `<input type="text">` than `<textarea>`), where paragraphs don't make sense.

**Comment by marijnh** [comment]
You appear to be using the default schema's `Doc` node type, which only allows block content.

That being said, with yesterday's master branch you're bound to run into issues even if you define your own `Doc` node that contains inline content. With the [changes](https://discuss.prosemirror.net/t/design-for-constrained-node-content/303/9) I just landed, things are likely better (though I still haven't tested this). Could you try checking out the current master branch, [defining](http://prosemirror.net/guide/schema.html) the schema you want, and seeing if it works as hoped?

Do report back here if something is still broken.

**Comment by theefer** [comment]
Apologies for the delay, I've finally managed to catch up with this. Thanks for your reply, and the new SchemaSpec format looks great!

I've tried again with ProseMirror 0.7.0 and the following code:

``` js
const customSchema = new Schema({
  nodes: {
    doc: {type: Doc, content: 'text*'},
    text: {type: Text}
  }
});

const contentDoc = customSchema.node('doc', {}, [
  customSchema.text('foo bar')
]);

const editorPlain = new ProseMirror({
  doc: contentDoc,
  schema: customSchema,
  place: document.getElementById('container')
});
```

But I still get the same error in `input.js` and `selection.js`, in both cases because the `range` object is `undefined`.

**Comment by marijnh** [comment]
You need to define your own `Doc` type that extends `Textblock` -- the default `Doc` is a block element, but elements that contain text should be textblocks.

(I hope to make this a little more automatic and less error-prone in 0.8.0)

**Comment by theefer** [comment]
Thanks for the quick reply!

I've tried your suggestion but I'm still getting the same error with:

``` js
export class TextDoc extends Textblock {}

const customSchema = new Schema({
  nodes: {
    doc: {type: TextDoc, content: 'text*'},
    text: {type: Text}
  }
});

const contentDoc = customSchema.node('doc', {}, [
  customSchema.text('foo bar')
]);
```

Unless I misunderstood you?

**Comment by marijnh** [comment]
I just set up (on the current master branch code) a simple test where I used the following schema:

```
const schema = new Schema({
  nodes: {
    doc: {type: Doc, content: "inline<_>*"},

    text: {type: Text, group: "inline"},
    hard_break: {type: HardBreak, group: "inline"}
  },

  marks: {
    strong: StrongMark,
  }
})
```

I ran across a bunch of issues, which I fixed in 0c7aff3 , after which I could work with the editor without problems. It's not unlikely that there are a bunch of things I didn't catch yet, but the basics should be fine now. Could you try it and open new issues for problems you run into?
